examples/grpc/main.go

```diff
@@ -0,0 +1,247 @@
+	config := provider.Config{}
+	bytes, err := os.ReadFile(ProviderConfigPath)
+	if err != nil {
+		panic(err)
+	}
+	err = json.Unmarshal(bytes, &config)
+	if err != nil {
+		panic(err)
+	}
+	client := grpc.NewGRPCClient(config, log)
+	err = client.Start(context.TODO())
+	if err != nil {
+		panic(err)
+	}
+	err = client.ProviderInit(context.TODO())
+	if err != nil {
+		panic(err)
+	}
+	ruleSets, err := ReadRulesets(RulesPath)
+	if err != nil {
+		return
+	}
+	results := []ResultRuleset{}
+	for _, rs := range ruleSets {
+		result, eErr := EvaluateRuleset(client, rs)
+		if eErr != nil {
+			return
+		}
+		results = append(results, result)
+	}
+
+	dump, _ := json.Marshal(results)
+	fmt.Printf("%s\n", dump)
+	client.Stop()
+}
+
+func EvaluateRuleset(svc provider.ServiceClient, rs RuleSet) (result ResultRuleset, err error) {
+	result.Name = rs.Name
+	result.Description = rs.Description
+	result.Violations = make(map[string]Violation)
+	result.Errors = make(map[string]string)
+	for _, r := range rs.Rules {
+		var resp provider.ProviderEvaluateResponse
+		var bytes []byte
+		if condInfo, ok := r.When["k8s.rego_expr"]; ok {
+			bytes, err = json.Marshal(condInfo)
+			if err != nil {
+				return
+			}
+			resp, err = svc.Evaluate(context.TODO(), "rego_expr", bytes)
+		} else if condInfo, ok = r.When["k8s.rego_module"]; ok {
+			bytes, err = json.Marshal(condInfo)
+			if err != nil {
+				return
+			}
+			resp, err = svc.Evaluate(context.TODO(), "rego_module", bytes)
+		}
+		if err != nil {
+			result.Errors[r.RuleID] = err.Error()
+			continue
+		}
+		if !resp.Matched {
+			result.Unmatched = append(result.Unmatched, r.RuleID)
+			continue
+		}
+		category := liboutput.Category(r.Category)
+		v := Violation{
+			Description: r.Message,
+			Category:    &category,
+			Labels:      nil,
+			Incidents:   nil,
+			Links:       nil,
+			Extras:      nil,
+			Effort:      &r.Effort,
+		}
+		for _, incidentContext := range resp.Incidents {
+			incident := liboutput.Incident{
+				URI:       incidentContext.FileURI,
+				Message:   r.Message,
+				Variables: incidentContext.Variables,
+			}
+			v.Incidents = append(v.Incidents, incident)
+		}
+		result.Violations[r.RuleID] = v
+	}
+
+	return
+}
+
+type ResultRuleset struct {
+	Name        string               `yaml:"name,omitempty" json:"name,omitempty"`
+	Description string               `yaml:"description,omitempty" json:"description,omitempty"`
+	Tags        []string             `yaml:"tags,omitempty" json:"tags,omitempty"`
+	Violations  map[string]Violation `yaml:"violations,omitempty" json:"violations,omitempty"`
+	Errors      map[string]string    `yaml:"errors,omitempty" json:"errors,omitempty"`
+	Unmatched   []string             `yaml:"unmatched,omitempty" json:"unmatched,omitempty"`
+	Skipped     []string             `yaml:"skipped,omitempty" json:"skipped,omitempty"`
+	Labels      []string             `yaml:"labels,omitempty" json:"labels,omitempty"`
+}
+
+type Violation struct {
```

Comment:
Could also import from https://github.com/konveyor/analyzer-lsp/blob/main/output/v1/konveyor/violations.go#L97

Reply:
I tried to import the Ruleset and Violations from there originally, but marshaling the structures to json/yaml was causing problems. The yaml encoder was recursing until it blew up the stack, and the json encoder was failing due to a type it couldn't understand.